Add a scheduled job to purge historical orders

putrenkov opened the pull request:
**A Brief Overview**
Add the ability for historical Orders and associated Customers to be purged from the system.

**Link to QA issue**
BroadleafCommerce/QA#3926

lMysticl committed:
BroadleafCommerce/QA-3926
Add the ability for historical Orders and associated Customers to be purged from the system.

Commit:
BroadleafCommerce/QA#3926
batch update

Commit:
BroadleafCommerce/QA#3926
added update statements in case of self-reference
added sorting by depth so delete order should be good now

Commit:
BroadleafCommerce/QA#3926
sql to be compatible with MySql and Postgres

Commit:
BroadleafCommerce/QA#3926
sql to be compatible with MySql and Postgres

lMysticl committed:
BroadleafCommerce/QA#3926
Add properties start and end date for purge. \
Add save protection for unexpected purge order history

Commit:
BroadleafCommerce/QA#3926
ManyToMany parsing added
JoinTable xrefs moved up, so executed earlier

Commit:
BroadleafCommerce/QA#3926
ManyToMany parsing added
JoinTable xrefs moved up, so executed earlier

Commit:
BroadleafCommerce/QA#3926
ManyToMany parsing added
JoinTable xrefs moved up, so executed earlier

Commit:
BroadleafCommerce/QA#3926
ManyToMany parsing added
JoinTable xrefs moved up, so executed earlier

Commit:
BroadleafCommerce/QA#3926
java docs

riteshadhikari17 commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/util/service/DeleteStatementGeneratorImpl.java:
```diff
@@ -0,0 +1,389 @@
+ * table name and path to that tabe from the root table, so we can build sql to delete dependencies in the right order.
+ * It knows how to deal with @Embedded, bidirectional @OneToMany, @OneToMany through @JoinTable
+ */
+@Service("blcDeleteStatementGenerator")
```
I think this should be changed to "blDeleteStatementGenerator". There is extra 'c'

riteshadhikari17 commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/util/service/ResourcePurgeServiceImpl.java:
```diff
@@ -102,6 +110,15 @@
     @Resource(name = "blCustomerService")
     protected CustomerService customerService;
 
+    @Resource(name = "blcDeleteStatementGenerator")
```
As mentioned above, change from "blcDeleteStatementGenerator" to "blDeleteStatementGenerator"

riteshadhikari17 commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/util/service/ResourcePurgeServiceImpl.java:
```diff
@@ -136,6 +153,69 @@ public void purgeCarts(final Map<String, String> config) {
     }
 
+    @Override
+    public void purgeHistory(Class<?> rootType, String rootTypeIdValue, Map<String, DeleteStatementGeneratorImpl.PathElement> depends) {
```
purgeHistory sounds little vague. Maybe "purgeOrderHistory" or "purgeHistoricOrders"

riteshadhikari17 commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/util/service/ResourcePurgeServiceImpl.java:
```diff
@@ -136,6 +153,69 @@ public void purgeCarts(final Map<String, String> config) {
+
+        String enablePurge = env.getProperty("enable.purge");
+
+        if (!Boolean.parseBoolean(enablePurge)) {
```
Again very vague. "enable.purge.order.history" or "enable.purge.historic.orders"

riteshadhikari17 commented on core/broadleaf-framework/src/main/resources/config/bc/fw/common.properties:
```diff
@@ -112,3 +112,9 @@ org.broadleafcommerce.core.util.lock.DistributedLock.canParticipate=true
+date.str=2011.11.19
+
+#Save protection from unexpected purge order history
+enable.purge=false
```
Same as above

lMysticl committed:
BroadleafCommerce/QA-3926
Rename for more clarification

putrenkov commented:
@riteshadhikari17 I pushed changes

Commit:
BroadleafCommerce/QA#3926
purge missing tables
support multiple dependency on table

riteshadhikari17 committed:
Renaming properties to be more explicit

Commit:
BroadleafCommerce/QA#3926
moved some sql statements up

riteshadhikari17 committed:
Added a statement to indicate end of process

Commit:
BroadleafCommerce/QA#3926
now cleared:
BLC_CREDIT_ACCOUNT_EVENT
BLC_CREDIT_ACCOUNT_EVENT_DTL
BLC_GIFTWRAP_ORDER_ITEM
BLC_GIFT_CARD_ACCOUNT

blc_credit_account - in case if now records for that account in blc_credit_account_event

Commit:
BroadleafCommerce/QA#3926
debug removed

Commit:
BroadleafCommerce/QA#3926
introduced extension handler, and moved some deletes to oms, quote, account credit modules

riteshadhikari17 commented:
@oleksiimiroshnyk It is still giving table not found errors. Looks like there are more references to credit_account tables
```
            creditAccountEventDependencies.add(new DeleteStatementGeneratorImpl.PathElement("BLC_CREDIT_ACCOUNT_EVENT_DTL", "CREDIT_ACCOUNT_EVENT_ID", "CREDIT_ACCOUNT_EVENT_ID"));
            creditAccountEventDependencies.add(new DeleteStatementGeneratorImpl.PathElement("BLC_CREDIT_ACCOUNT_EVENT_DTL", "CREDIT_ACCOUNT_EVENT_ID", "DEBIT_ACCOUNT_EVENT_ID"));
```

Commit:
BroadleafCommerce/QA#3926
introduced extension handler, and moved some deletes to oms, quote, account credit modules

Commit:
BroadleafCommerce/QA#3926
introduced extension handler, and moved some deletes to oms, quote, account credit modules

Commit:
BroadleafCommerce/QA#3926
introduced extension handler, and moved some deletes to oms, quote, account credit modules

Commit:
BroadleafCommerce/QA#3926
introduced extension handler, and moved some deletes to oms, quote, account credit modules

troian88 committed:
Merge branch 'develop-6.0.x' into QA-3926-historical-purge

troian88 committed:
Add Days count and Limit to the historical purge parameters (#2495)

* Added days and limit parameters

* Add Variable names

* Refactoring

troian88 committed:
Removed unused historical purge properties

riteshadhikari17 commented on core/broadleaf-framework/src/main/java/org/broadleafcommerce/core/util/service/ResourcePurgeServiceImpl.java:
```diff
@@ -136,6 +156,71 @@ public void purgeCarts(final Map<String, String> config) {
+        }
+
+        Integer daysCount = config.get(OLDER_THAN_DAYS.toString());
+        Integer batchSize = config.get(BATCH_SIZE.toString());
```
This is batch size defined in line 94. Instead of static import, import whole class and do following
```
 Integer daysCount = config.get(PurgeOrderHistoryVariableNames.OLDER_THAN_DAYS.toString());
 Integer batchSize = config.get(PurgeOrderHistoryVariableNames.BATCH_SIZE.toString());
```

riteshadhikari17 committed:
Fixed the incorrect reference to batch size.

troian88 committed:
Changed select query for the orders days count

Commit:
BroadleafCommerce/QA#3926
Fix for updates before deletes found for blc_order_admin_assignment, blc_quote when deleting order.

riteshadhikari17 committed:
Set the query to ascending order so that the purge starts with the oldest order

riteshadhikari17 committed:
Merge branch 'develop-6.0.x' into QA-3926-historical-purge

riteshadhikari17 committed:
Remove commented dependencies for woodstox

danielcolgrove committed:
Added a debug statement to emit the delete statements for order purge.
